### DISCUSSION TITLE
Feature proposal (volunteer to implement): Readonly at Connection level

### DISCUSSION THREAD

**Discussion by rootselaar** [duckdb/duckdb#11612] [Ideas]
Currently DuckDB supports readonly mode at the DB level.

For our analytics application it would be great if the readonly property can be specified at the level of a DB Connection.

The reason for this is that in our application we have a server that:
- on the one hand receives updates with new 'raw data' coming in. These must be written to the database.
- on the other hand we have user queries coming in. These users are allowed a lot of freedom and may create temporary tables and are allowed to query all the data, but they should not (accidentally) interfere with each other, so they should not only be able to perform write access to the database.

So what I am proposing here, is not a full blown access control, but a (much) simpler restriction of mutability at the connection level. I understand from earlier discussions that ['More granular access control are on [the] long term roadmap'](https://github.com/duckdb/duckdb/discussions/10000) so I am guessing that this proposal may be in line with the roadmap (at least it's not an explicit non-goal as far as I can tell).

I would be happy to actually implement this for the DuckDB C++ code and make a pull request for this. 
As proof of concept I will also adapt the NodeJS connector, so that we have an end-to-end working application.

Below is how I propose to go about this.

Let me know what you think.

# Implementation details 

Here is a rough outline of the implementation details.

## Connection side

(1) Add a a public member `bool IsReadonly()`' to the `Connection` class. 
(2) Allow specifying whether a connection is readonly at construction time (not sure yet what the cleanest way would be, e.g., an enum or a boolean passed in at the constructor, passed as part of an options object or separately).

## Binder side

(1) When binding mutating statements check whether they are in with `IsReadonly()`. If not, throw a BinderExceptionBinderException. For example the Bind for the InsertStatement would look something like this:

```
BoundStatement Binder::Bind(InsertStatement &stmt) {
	BoundStatement result;
	result.names = {"Count"};
	result.types = {LogicalType::BIGINT};

	BindSchemaOrCatalog(stmt.catalog, stmt.schema);
	auto &table = Catalog::GetEntry<TableCatalogEntry>(context, stmt.catalog, stmt.schema, stmt.table);
	if (!table.temporary) {

		if(context.IsReadonly()) {

			throw BinderException("INSERT cannot be done on a non-temporary table with a READ_ONLY connection.");

		}

      ..

    }    
   ...
}
```
Also the binding operator for the other mutating statements should be adapted of course.

**Comment by rootselaar** [comment]
# Regarding the implementation details

After investigating a bit further I see that it is probably better (for _most functions_) to do the check for `IsReadonly()` in the function `ClientContext::CheckIfPreparedStatementIsExecutable`. This follows the existing pattern for the check at the DB level.  

However, we would still need to have additional checks for `ATTACH` and `DETACH` which we would also like to exclude. 

# Writing to files 
And then, there is the question of how to deal with writing to _files_. These are, strictly speaking, not part of the DB, but the intent of a read-only connection is to avoid any kind of persistent mutations on the 'side of the DB server'.

**Reply by Tishj** [reply]
> Writing to files

`enable_external_access` exists for that purpose
